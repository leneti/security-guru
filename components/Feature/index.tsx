export { default as Feature } from "./Feature";
export * from "./types";
