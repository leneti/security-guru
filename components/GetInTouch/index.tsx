export { default, type ContactForm } from "./GetInTouch";
