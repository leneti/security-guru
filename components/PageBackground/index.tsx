export { default, type PageBGProps } from "./PageBackground";
