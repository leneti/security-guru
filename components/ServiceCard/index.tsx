export { default as ServiceCard } from "./ServiceCard";
export * from "./types";
