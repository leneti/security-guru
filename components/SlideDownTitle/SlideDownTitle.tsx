import { useState, useEffect } from "react";
import { useRouter } from "next/router";
import {
  Title,
  Transition,
  type MantineTransition,
  Text,
  createStyles,
  clsx,
} from "@mantine/core";
import { useReducedMotion } from "@mantine/hooks";
import { flatLinks } from "@site/constants";
import { usePrevUrlContext } from "@site/contexts/prevUrl";
import { SlideDownTitleProps } from "./types";

const marginBottom = "1.5rem";

const scaleY: MantineTransition = {
  in: {
    opacity: 1,
    transform: "scaleY(1)",
    marginBottom,
    height: "100%",
  },
  out: {
    opacity: 0,
    transform: "scaleY(0)",
    marginBottom: 0,
    height: 0,
  },
  common: {
    transformOrigin: "top",
    overflow: "hidden",
  },
  transitionProperty: "transform, opacity, margin-bottom height",
};

const useStyles = createStyles(
  (
    theme,
    {
      noMotion,
      mounted,
      direction,
    }: { noMotion: boolean; mounted: boolean; direction: "right" | "left" }
  ) => ({
    title: {
      fontWeight: 900,
      marginTop: theme.spacing.xs,
      marginBottom,
      paddingBottom: theme.spacing.sm,
      paddingLeft: theme.spacing.xl,
      paddingRight: theme.spacing.xl,
    },

    horizontallyMovingTitle: {
      opacity: mounted || noMotion ? 1 : 0,
      ...(!noMotion
        ? {
            transform: `scaleX(${mounted ? 1 : 0.9})`,
            transitionProperty: "opacity transform",
            transitionDuration: ".2s",
            transformOrigin: `${direction} center 0px`,
          }
        : {}),
    },
  })
);

const WIPText = ({ wip }: Pick<SlideDownTitleProps, "wip">) =>
  wip ? (
    <Text align="center" mb="lg" c="dimmed">
      Puslapis ruo≈°iamas
    </Text>
  ) : null;

export default function SlideDownTitle(props: SlideDownTitleProps) {
  const { title, wip = false } = props;
  const noMotion = useReducedMotion();
  const [mounted, setMounted] = useState(false);
  const prevUrl = usePrevUrlContext();
  const currentUrl = useRouter().asPath;
  const direction =
    flatLinks.findIndex(({ url }) => url === currentUrl) >
    flatLinks.findIndex(({ url }) => url === prevUrl)
      ? "right"
      : "left";
  const { classes } = useStyles({ noMotion, mounted, direction });

  useEffect(() => setMounted(true), []);

  return prevUrl === "/" ? (
    <Transition mounted={mounted} transition={scaleY} duration={200}>
      {(styles) => (
        <>
          <Title align="center" style={styles} className={classes.title}>
            {title}
          </Title>
          <WIPText wip={wip} />
        </>
      )}
    </Transition>
  ) : (
    <>
      <Title
        align="center"
        className={clsx(classes.title, classes.horizontallyMovingTitle)}
      >
        {title}
      </Title>
      <WIPText wip={wip} />
    </>
  );
}
