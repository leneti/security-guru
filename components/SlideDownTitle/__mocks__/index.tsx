import type { SlideDownTitleProps } from "../types";

export function SlideDownTitle(props: SlideDownTitleProps) {
  const { title, wip } = props;
  return (
    <h1 className={`mock-slide-down-title${wip ? " WIP" : ""}`}>{title}</h1>
  );
}
