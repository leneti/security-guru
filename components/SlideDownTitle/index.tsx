export { default as SlideDownTitle } from "./SlideDownTitle";
export * from "./types";
