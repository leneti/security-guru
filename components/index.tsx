export { default as ContactInfo } from "./ContactInfo";
export { default as Footer } from "./Footer";
export { default as GetInTouch, type ContactForm } from "./GetInTouch";
export { default as Header } from "./Header";
export { default as Hero } from "./Hero";
export { default as Logo } from "./Logo";
export { default as PageTitle } from "./PageTitle";
export { default as PageBackground, type PageBGProps } from "./PageBackground";
export { default as SlideDownTitle } from "./SlideDownTitle";
