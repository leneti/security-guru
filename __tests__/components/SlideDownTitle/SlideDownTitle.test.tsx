import { render, screen, waitFor } from "@testing-library/react";
import { SlideDownTitle } from "@site/components/SlideDownTitle";

jest.mock("@site/contexts/prevUrl", () => ({
  __esModule: true,
  usePrevUrlContext: jest.fn(() => "/").mockName("usePrevUrlContext"),
}));
jest.mock("next/router", () => ({
  ...jest.requireActual("next/router"),
  useRouter: () => ({
    asPath: "/mock-page",
    events: {
      on: jest.fn(),
      off: jest.fn(),
    },
  }),
}));
jest.mock("@site/constants", () => ({
  flatLinks: [{ url: "/" }, { url: "/mock-page" }, { url: "/not-home-page" }],
}));

const mockTitle = "mock-title";

describe("SlideDownTitle", () => {
  it("renders title sliding from top", async () => {
    render(<SlideDownTitle title={mockTitle} />);

    const heading = screen.getByRole("heading", { name: mockTitle });
    expect(heading).toBeInTheDocument();
    expect(heading).toHaveProperty("style.transition-duration", "200ms");

    await waitFor(() => expect(heading).toHaveProperty("style.opacity", "1"));
    await waitFor(() =>
      expect(heading).toHaveProperty("style.transform", "scaleY(1)")
    );
  });

  it.each([
    { from: "/not-home-page", to: "/mock-page", side: "right" },
    { from: "/mock-page", to: "/not-home-page", side: "left" },
  ])("renders title sliding from $side", ({ from, to }) => {
    jest
      .spyOn(require("@site/contexts/prevUrl"), "usePrevUrlContext")
      .mockReturnValue(from);

    jest.spyOn(require("next/router"), "useRouter").mockImplementation(() => ({
      asPath: to,
      events: {
        on: jest.fn(),
        off: jest.fn(),
      },
    }));

    const { container } = render(<SlideDownTitle title={mockTitle} />);

    expect(
      screen.getByRole("heading", { name: mockTitle })
    ).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("renders WIP text", () => {
    render(<SlideDownTitle title={mockTitle} wip />);
    expect(screen.getByText("Puslapis ruoÅ¡iamas")).toBeInTheDocument();
  });

  it("renders title with no transition", () => {
    jest
      .spyOn(jest.requireActual("@mantine/hooks"), "useReducedMotion")
      .mockReturnValue(true);

    const { container } = render(<SlideDownTitle title={mockTitle} />);

    expect(container).toMatchSnapshot();
  });
});
