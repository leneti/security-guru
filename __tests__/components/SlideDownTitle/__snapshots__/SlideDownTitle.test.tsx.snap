// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlideDownTitle renders title sliding from left 1`] = `
<div>
  <h1
    class="mantine-Text-root mantine-Title-root mantine-q55mz5"
  >
    mock-title
  </h1>
</div>
`;

exports[`SlideDownTitle renders title sliding from right 1`] = `
<div>
  <h1
    class="mantine-Text-root mantine-Title-root mantine-106pfj9"
  >
    mock-title
  </h1>
</div>
`;

exports[`SlideDownTitle renders title with no transition 1`] = `
<div>
  <h1
    class="mantine-Text-root mantine-Title-root mantine-cx5884"
  >
    mock-title
  </h1>
</div>
`;
